### PR TITLE
Document how to add a new metric type in the code

### PR DESCRIPTION
**Pull request by badboy**
### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - `make test` runs without emitting any warnings
  - `make lint` runs without emitting any errors
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry or an explanation of why it does not need one
  - Any breaking changes to language binding APIs are noted explicitly

### CODE REVIEW

File: docs/dev/core/new-metric-type.md [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    COUNTER_METRICS.insert_with_log(|| {
+        let send_in_pings = unsafe { from_raw_string_array(send_in_pings, send_in_pings_len) };
+        let lifetime = Lifetime::try_from(lifetime)
+            .map_err(|_| {
+                log::error!(
+                    "[Counter] Failed to convert from lifetime value {}",
+                    lifetime
+                );
+            })
+            .unwrap_or(Lifetime::Ping);
```
Comment (mdboom) [inline docs/dev/core/new-metric-type.md]: An aside (not for this PR): Is there a way to encapsulate this to not duplicate it in every metric type constructor?

File: docs/dev/core/new-metric-type.md [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        COUNTER_METRICS.call_infallible(metric_id, |metric| {
+            metric.add(glean, amount);
+        })
+    })
+}
+```
+
+### Platform-part (Kotlin)
+
+The platform-specific FFI wrapper needs the definitons of these new functions.
```
Comment (mdboom) [inline docs/dev/core/new-metric-type.md]:
```suggestion
The platform-specific FFI wrapper needs the definitions of these new functions.
```

File: docs/dev/core/new-metric-type.md [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        match self {
+            // ...
+            Metric::Counter(c) => json!(c),
+        }
+    }
+}
+```
+
+For more complex serialization consider implementing serialization logic as a function returning a [`serde_json::Value`](https://docs.rs/serde_json/*/serde_json/enum.Value.html)
+or another object that can be serialized.
```
Comment (mdboom) [inline docs/dev/core/new-metric-type.md]: I feel like we'll want a concrete example of this, but we can defer that until we actually have one in the code base.  In particular, documenting the convention that the storage struct (as distinct from the metric definition struct) lives in the same module as the metric type etc., etc...

File: docs/dev/core/new-metric-type.md [hunk 4]
```diff
@@ -0,0 +1,212 @@
+# Adding a new metric type
+
+Data in Glean is stored in so called metrics.
```
Comment (mdboom) [inline docs/dev/core/new-metric-type.md]:
```suggestion
Data in Glean is stored in so-called metrics.
```

File: docs/dev/core/new-metric-type.md [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+You can find the full list of implemented metric types [in the user overview](/user/metrics/index.md).
+
+Adding a new metric type involves defining the metric type's API, its persisted and in-memory storage as well as its serialization into the ping payload.
+
+## The metric type's API
+
+A metric type implementation is defined in its own file under `glean-core/src/metrics/`, e.g. `glean-core/src/metrics/counter.rs` for a [Counter](/user/metrics/counter.md).
+
+Start by defining a structure:
```
Comment (mdboom) [inline docs/dev/core/new-metric-type.md]: ... to hold the metric's metadata.

(Or something, to make it clear that this structure is holding the metric definition, not the data for each recording of the metric)...

File: docs/dev/core/new-metric-type.md (hunk 3 above)
Comment (badboy) [reply docs/dev/core/new-metric-type.md]: Yeah, I was thinking about that but then I was half-way into implementing Datetime and stopped.
I'll file a bug so we replace this once we have a more concrete example.

File: docs/dev/core/new-metric-type.md (hunk 1 above)
Comment (badboy) [reply docs/dev/core/new-metric-type.md]: Yes. Actually we can use `from` and default to `Lifetime::Ping`. I'll change it here and add another PR changing the code.

File: docs/dev/core/new-metric-type.md (hunk 3 above)
Comment (badboy) [reply docs/dev/core/new-metric-type.md]: https://bugzilla.mozilla.org/show_bug.cgi?id=1552774

File: docs/dev/core/new-metric-type.md [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+    category: FfiStr,
+    name: FfiStr,
+    send_in_pings: RawStringArray,
+    send_in_pings_len: i32,
+    lifetime: i32,
+    disabled: u8,
+) -> u64 {
+    COUNTER_METRICS.insert_with_log(|| {
+        let send_in_pings = unsafe { from_raw_string_array(send_in_pings, send_in_pings_len) };
+        let lifetime = Lifetime::try_from(lifetime)?;
```
Comment (Dexterp37) [inline docs/dev/core/new-metric-type.md]: Looks much nicer :)

Comment (badboy) [reply docs/dev/core/new-metric-type.md]: I agree.

File: docs/dev/core/new-metric-type.md (hunk 3 above)
Comment (mdboom) [reply docs/dev/core/new-metric-type.md]: Yeah, sounds good.  I think we don't fully know all the details on this until we implement the first one anyway...

**Pull request merged**